File: src/Heap/MaxHeap.java
package Heap;

public class MaxHeap {
    private int[] heap;
    private int size;
    private int max;

    public MaxHeap(int max){
        this.size = 0;
        this.max = max;
        heap = new int[this.max];
    }

    public void insert(int node){

        if (size >= max){
            System.out.println("Insertion not possible");
            return;
        }

        heap[size] = node;
        int currentNodeIndex = size;

        while (heap[currentNodeIndex] > heap[parentNode(currentNodeIndex)]){
            swap(currentNodeIndex, parentNode(currentNodeIndex));
            currentNodeIndex = parentNode(currentNodeIndex);
        }
        size++;
    }

    private void maxHeapify(int currentIdx){
        if (isLeafNode(currentIdx)) return;

        int left = leftChild(currentIdx);
        int right = rightChild(currentIdx);

        if (heap[currentIdx] < heap[left] || heap[currentIdx] < heap[right]){
            if (heap[left] > heap[right]){
                swap(currentIdx, left);
                maxHeapify(left);
            }
            else{
                swap(currentIdx, right);
                maxHeapify(right);
            }
        }
    }

    public int removeElement(){
        int remElement = heap[0];
        heap[0] = heap[--size];
        maxHeapify(0);
        return remElement;
    }

    private int parentNode(int index){return (index-1)/2;}

    private int leftChild(int index){return 2*index + 1;}

    private int rightChild(int index){return 2*index + 2;}

    public void print(){
        for (int i = 0; i < size / 2; i++) {
            System.out.print("Parent : " + heap[i]);

            if (leftChild(i) < size)
                System.out.print(" Left Child :" + heap[leftChild(i)]);

            if (rightChild(i) < size)
                System.out.println(" Right Child: " + heap[rightChild(i)]);

            System.out.println();
        }
    }

    private boolean isLeafNode(int index){
        return index > (size / 2) && index <= size;
    }

    public void swap(int index1, int index2){
        int temp = heap[index1];
        heap[index1] = heap[index2];
        heap[index2] = temp;
    }
    public static void main(String[] args) {
        MaxHeap maxHeap = new MaxHeap(11);

        maxHeap.insert(1);
        maxHeap.insert(3);
        maxHeap.insert(5);
        maxHeap.insert(4);
        maxHeap.insert(6);
        maxHeap.insert(13);
        maxHeap.insert(10);
        maxHeap.insert(9);
        maxHeap.insert(8);
        maxHeap.insert(15);
        maxHeap.insert(17);



        maxHeap.print();

        System.out.println("Remove max :" + maxHeap.removeElement());
        maxHeap.print();
    }
}

File: Heap/js/MaxHeap.js
import { Comparator } from "../../utils/Comparator.js";

class MaxHeap {
  constructor() {
    this.heapArray = [];
    this.comparator = new Comparator();
  }

  getLeftChildIndex(parentIndex) {
    return 2 * parentIndex + 1;
  }

  getRightChildIndex(parentIndex) {
    return 2 * parentIndex + 2;
  }

  getParentIndex(childIndex) {
    return Math.floor((childIndex - 1) / 2);
  }

  hasLeftChild(parentIndex) {
    return this.getLeftChildIndex(parentIndex) < this.heapArray.length;
  }

  hasRightChild(parentIndex) {
    return this.getRightChildIndex(parentIndex) < this.heapArray.length;
  }

  leftChild(parentIndex) {
    return this.heapArray[this.getLeftChildIndex(parentIndex)];
  }

  rightChild(parentIndex) {
    return this.heapArray[this.getRightChildIndex(parentIndex)];
  }

  parent(childIndex) {
    return this.heapArray[this.getParentIndex(childIndex)];
  }

  swap(firstIndex, secondIndex) {
    const temp = this.heapArray[firstIndex];
    this.heapArray[firstIndex] = this.heapArray[secondIndex];
    this.heapArray[secondIndex] = temp;
  }

  peek() {
    if (this.heapArray.length === 0) {
      return null;
    }

    return this.heapArray[0];
  }

  add(nodeValue) {
    this.heapArray.push(nodeValue);
    this.heapifyUp();
  }

  poll() {
    if (this.heapArray.length === 0) {
      return null;
    }

    if (this.heapArray.length === 1) {
      return this.heapArray.pop();
    }

    const item = this.heapArray[0];

    // replace last element from the end to head
    this.heapArray[0] = this.heapArray.pop();
    this.heapifyDown();

    return item;
  }

  isEmpty() {
    return this.heapArray.length === 0;
  }

  toString() {
    return this.heapArray.join(",");
  }

  heapifyUp(startingNodeIndex = 0) {
    let currentNodeIndex = startingNodeIndex || this.heapArray.length - 1;

    while (currentNodeIndex > 0) {
      const parentIndex = this.getParentIndex(currentNodeIndex);
      if (this.heapArray[currentNodeIndex] <= this.heapArray[parentIndex]) {
        break;
      }
      this.swap(parentIndex, currentNodeIndex);
      currentNodeIndex = parentIndex;
    }
  }

  heapifyDown(startingNodeIndex = 0) {
    let currentNodeIndex = startingNodeIndex;

    while (
      (this.hasLeftChild(currentNodeIndex) &&
        this.heapArray[currentNodeIndex] < this.leftChild(currentNodeIndex)) ||
      (this.hasRightChild(currentNodeIndex) &&
        this.heapArray[currentNodeIndex] < this.rightChild(currentNodeIndex))
    ) {
      const hasLeftChild = this.hasLeftChild(currentNodeIndex);
      const hasRightChild = this.hasRightChild(currentNodeIndex);
      const leftChildIndex = this.getLeftChildIndex(currentNodeIndex);

      if (hasLeftChild && hasRightChild) {
        const left = this.leftChild(currentNodeIndex);
        const right = this.rightChild(currentNodeIndex);
        const rightChildIndex = this.getRightChildIndex(currentNodeIndex);

        if (left >= right) {
          this.swap(currentNodeIndex, leftChildIndex);
          currentNodeIndex = leftChildIndex;
        } else {
          this.swap(currentNodeIndex, rightChildIndex);
          currentNodeIndex = rightChildIndex;
        }
      } else if (hasLeftChild) {
        this.swap(currentNodeIndex, leftChildIndex);
        currentNodeIndex = leftChildIndex;
      }
    }
  }

  find(node, compare = this.comparator) {
    const foundIndices = [];

    for (let index = 0; index < this.heapArray.length; index++) {
      const element = this.heapArray[index];
      if (compare.equal(element, node)) {
        foundIndices.push(index);
      }
    }

    return foundIndices;
  }

  remove(removeNode) {
    let totalNodeFound = this.find(removeNode).length;

    while (totalNodeFound-- > 0) {
      const removeNodeIndex = this.find(removeNode).pop();

      if (removeNodeIndex === this.heapArray.length - 1) {
        this.heapArray.pop();
        continue;
      }

      this.heapArray[removeNodeIndex] = this.heapArray.pop();
      const parent = this.parent(removeNodeIndex);
      const hasLeftChild = this.hasLeftChild(removeNodeIndex);

      if (
        hasLeftChild &&
        (!parent ||
          this.comparator.greaterThanOrEqual(
            parent,
            this.heapArray[removeNodeIndex]
          ))
      ) {
        this.heapifyDown(removeNodeIndex);
      } else {
        this.heapifyUp(removeNodeIndex);
      }
    }

    return this;
  }
}

export default MaxHeap;
